### PR TITLE
testing pypi release test

### PR DESCRIPTION
**Pull request by binaryaaron**
Adding in the proper GH action file to handle both 

- pypi releases (test and prod)
- GH releases (draft and real)

The test path should get triggered on PRs to main; the others will only get triggered if we add a tag.

### CODE REVIEW

**Comment by sfriedowitz** [comment]
Why are these jobs running as part of the PR checks? It seems like the triggers might be incorrect?

File: .github/workflows/publish.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+      - name: Publish dist to real pypi
+        env:
+            POETRY_PYPI_TOKEN: ${{ secrets.PYPI_KEY }}
+        if: endsWith(github.event.base_ref, 'main') == true
+        run: |
+            source "$HOME/venv/bin/activate"
+            poetry publish 
+
+    github-release:
+        name: prepare release
```
Comment (sfriedowitz) [inline .github/workflows/publish.yaml:65]:
```suggestion
        name: Prepare release
```

**Comment by binaryaaron** [comment]
> Why are these jobs running as part of the PR checks? It seems like the triggers might be incorrect if this is supposed to only run on publishes to `main` or GH releases?

Only the draft ones are done when there is a tag. so this should _not_ run if you make a tag-less PR against main. 

can basically have a branch like "dev-0.2.0" and when it's ready to release, make a PR against main with a tag `v0.2.0`, which will then trigger this. after actual changes are landed, the non-draft version should run from main, and push the wheels to real pypi.

**Comment by veekaybee** [comment]
Could you add some instructions in contributing around this and how to add a tag, i.e. is this a git tag? @binaryaaron 

>can basically have a branch like "dev-0.2.0" and when it's ready to release, make a PR against main with a tag v0.2.0, which will then trigger this. after actual changes are landed, the non-draft version should run from main, and push the wheels to real pypi.

File: .github/workflows/publish.yaml [hunk 2]
```diff
@@ -0,0 +1,94 @@
+# Build and publish to PyPI
```
Comment (sfriedowitz) [inline .github/workflows/publish.yaml:1]:
```suggestion
name: Build and publish to PyPI
```

Should name the workflow a as a whole, or else it displays weird in the PR checks

**Comment by sfriedowitz** [comment]
I'm a bit confused by the release process here, my thought was that we would create a GH Release with a newly created tag based off the main branch, and that would trigger the publishing actions to real pypi.

I'm fine to go with this to give this system a shot, though.

**Comment by binaryaaron** [comment]
> I'm a bit confused by the release process here, my thought was that we would create a GH Release with a newly created tag based off the main branch, and that would trigger the publishing actions to real pypi.
> 
> I'm fine to go with this to give this system a shot, though. We can always tweak down the line if the triggers are not exactly as we'd expect.
> 
> Edit: This is the release trigger I was thinking of (https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release)

that'd be fine too! this is more automation than that; if it doesn't work happy to simplify it / create a better overall structure for using GHA. 

I'll add docs, @veekaybee , ty for the reminder.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -115,3 +116,44 @@ When you're ready, run:
 ```
 poetry publish --build
 ```
+
+## Publishing  (automated)
+
+`.github/workflows/publish.yaml' contains the GitHub Action used to do releases and push wheels to PyPI.
```
Comment (sfriedowitz) [inline CONTRIBUTING.md:122]:
```suggestion
`.github/workflows/publish.yaml` contains the GitHub Action used to do releases and push wheels to PyPI.
```

Ticks were off, so was not displaying code style

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -115,3 +116,44 @@ When you're ready, run:
...
+
+There are two subcases - draft/dev and the 'real' release.
```
Comment (sfriedowitz) [inline CONTRIBUTING.md:124]:
```suggestion
There are two cases - 'draft' and the 'real' releases.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -115,3 +116,44 @@ When you're ready, run:
...
+
+
+### Draft / Test releases
+
+If a contributed opens a PR to `main` with a git version tag (e.g., `vX.Y.Z`) , the draft process will start which does the following:
```
Comment (sfriedowitz) [inline CONTRIBUTING.md:129]:
```suggestion
If a contributer opens a PR to `main` with a git version tag (e.g., `vX.Y.Z`) , the draft process will start which does the following:
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -115,3 +116,44 @@ When you're ready, run:
...
+
+- upload a version to test PyPI 
+- make a draft release on github.
+
+
+this is to ensure the full process works before merging to main.
```
Comment (sfriedowitz) [inline CONTRIBUTING.md:135]:
```suggestion
This is to ensure the full process works before merging to `main`.
```

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -115,3 +116,44 @@ When you're ready, run:
...
 ```
+
+## Publishing  (automated)
+
+`.github/workflows/publish.yaml' contains the GitHub Action used to do releases and push wheels to PyPI.
+
+There are two subcases - draft/dev and the 'real' release.
+
+
+### Draft / Test releases
```
Comment (sfriedowitz) [inline CONTRIBUTING.md:127]:
```suggestion
### Draft releases
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -115,3 +116,44 @@ When you're ready, run:
...
+If a contributed opens a PR to `main` with a git version tag (e.g., `vX.Y.Z`) , the draft process will start which does the following:
+
+- upload a version to test PyPI 
+- make a draft release on github.
+
+
+this is to ensure the full process works before merging to main.
+
+
+### Real release
```
Comment (sfriedowitz) [inline CONTRIBUTING.md:138]:
```suggestion
### Real releases
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -115,3 +116,44 @@ When you're ready, run:
...
+- push your changes to upstream 
+- open a PR to main
+- verify that the project builds
+- land PR
+
+
+
+
+
+
```
Comment (sfriedowitz) [inline CONTRIBUTING.md:159]:
```suggestion
```

**Review by sfriedowitz — APPROVED** [review]
committable suggestions for consistency in formatting of contributing guide

**Comment by binaryaaron** [comment]
I'll add some followups after this one; i think using `labels` might be a little easier both from a user perspective and event filtering perspective.

**Pull request merged**